File: internal/plugins/jest/rules/no_jasmine_globals/no_jasmine_globals_test.go
package no_jasmine_globals_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_jasmine_globals"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoJasmineGlobalsRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_jasmine_globals.NoJasmineGlobalsRule,
		[]rule_tester.ValidTestCase{
			{Code: `jest.spyOn()`},
			{Code: `jest.fn()`},
			{Code: `expect.extend()`},
			{Code: `expect.any()`},
			{Code: `it("foo", function () {})`},
			{Code: `test("foo", function () {})`},
			{Code: `foo()`},
			{Code: `require("foo")("bar")`},
			{Code: `(function(){})()`},
			{Code: `function callback(fail) { fail() }`},
			{Code: `var spyOn = require("actions"); spyOn("foo")`},
			{Code: `function callback(pending) { pending() }`},
			{Code: `jasmine.DEFAULT_TIMEOUT_INTERVAL += 1000;`},
			{Code: `jasmine["DEFAULT_TIMEOUT_INTERVAL"] += 1000;`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code:   `spyOn(some, "object")`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalGlobal",
					Message:   "Illegal usage of global `spyOn`, prefer `jest.spyOn`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `spyOnProperty(some, "object")`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalGlobal",
					Message:   "Illegal usage of global `spyOnProperty`, prefer `jest.spyOn`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `fail()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalFail",
					Message:   "Illegal usage of `fail`, prefer throwing an error, or the `done.fail` callback",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `pending()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalPending",
					Message:   "Illegal usage of `pending`, prefer explicitly skipping a test using `test.skip`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.DEFAULT_TIMEOUT_INTERVAL = 5000;`,
				Output: []string{`jest.setTimeout(5000);`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.DEFAULT_TIMEOUT_INTERVAL = function() {}`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine["DEFAULT_TIMEOUT_INTERVAL"] = 5000;`,
				Output: []string{`jest.setTimeout(5000);`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine["DEFAULT_TIMEOUT_INTERVAL"] = function() {}`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.addMatchers(matchers)`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.addMatchers`, prefer `expect.extend`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.createSpy()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.createSpy`, prefer `jest.fn`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.any()`,
				Output: []string{`expect.any()`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.any`, prefer `expect.any`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.anything()`,
				Output: []string{`expect.anything()`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.anything`, prefer `expect.anything`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.arrayContaining()`,
				Output: []string{`expect.arrayContaining()`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.arrayContaining`, prefer `expect.arrayContaining`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.objectContaining()`,
				Output: []string{`expect.objectContaining()`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.objectContaining`, prefer `expect.objectContaining`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.stringMatching()`,
				Output: []string{`expect.stringMatching()`},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalMethod",
					Message:   "Illegal usage of `jasmine.stringMatching`, prefer `expect.stringMatching`",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.getEnv()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.empty()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.falsy()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.truthy()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.arrayWithExactContents()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.clock()`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine.MAX_PRETTY_PRINT_ARRAY_LENGTH = 42`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
			{
				Code:   `jasmine["MAX_PRETTY_PRINT_ARRAY_LENGTH"] = 42`,
				Output: []string{},
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "illegalJasmine",
					Message:   "Illegal usage of jasmine global",
					Column:    1,
					Line:      1,
				}},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-jasmine-globals.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-jasmine-globals', {} as never, {
  valid: [
    { code: 'jest.spyOn()' },
    { code: 'jest.fn()' },
    { code: 'expect.extend()' },
    { code: 'expect.any()' },
    { code: 'it("foo", function () {})' },
    { code: 'test("foo", function () {})' },
    { code: 'foo()' },
    { code: `require('foo')('bar')` },
    { code: '(function(){})()' },
    { code: 'function callback(fail) { fail() }' },
    { code: 'var spyOn = require("actions"); spyOn("foo")' },
    { code: 'function callback(pending) { pending() }' },
    { code: 'function callback(jasmine) { jasmine.any() }' },
    { code: 'const jasmine = foo;' },
    { code: '(this as any).jasmine.foo()' },
    { code: '(this as any).jasmine.any()' },
    { code: 'jasmine[1].any()' },
  ],
  invalid: [
    {
      code: 'spyOn(some, "object")',
      output: null,
      errors: [
        {
          messageId: 'illegalGlobal',
          data: { global: 'spyOn', replacement: 'jest.spyOn' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'spyOnProperty(some, "object")',
      output: null,
      errors: [
        {
          messageId: 'illegalGlobal',
          data: { global: 'spyOnProperty', replacement: 'jest.spyOn' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'fail()',
      output: null,
      errors: [{ messageId: 'illegalFail', column: 1, line: 1 }],
    },
    {
      code: 'pending()',
      output: null,
      errors: [{ messageId: 'illegalPending', column: 1, line: 1 }],
    },
    {
      code: 'jasmine',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'const value = jasmine;',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 15, line: 1 }],
    },
    {
      code: 'jasmine.DEFAULT_TIMEOUT_INTERVAL = 5000;',
      output: 'jest.setTimeout(5000);',
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.DEFAULT_TIMEOUT_INTERVAL = function() {}',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.DEFAULT_TIMEOUT_INTERVAL += 1000;',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine["DEFAULT_TIMEOUT_INTERVAL"] = 5000;',
      output: 'jest.setTimeout(5000);',
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine["DEFAULT_TIMEOUT_INTERVAL"] = function() {}',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine["DEFAULT_TIMEOUT_INTERVAL"] += 1000;',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.addMatchers(matchers)',
      output: null,
      errors: [
        {
          messageId: 'illegalMethod',
          data: { method: 'jasmine.addMatchers', replacement: 'expect.extend' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.createSpy()',
      output: null,
      errors: [
        {
          messageId: 'illegalMethod',
          data: { method: 'jasmine.createSpy', replacement: 'jest.fn' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.any()',
      output: 'expect.any()',
      errors: [
        {
          messageId: 'illegalMethod',
          data: { method: 'jasmine.any', replacement: 'expect.any' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.anything()',
      output: 'expect.anything()',
      errors: [
        {
          messageId: 'illegalMethod',
          data: { method: 'jasmine.anything', replacement: 'expect.anything' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.arrayContaining()',
      output: 'expect.arrayContaining()',
      errors: [
        {
          messageId: 'illegalMethod',
          data: {
            method: 'jasmine.arrayContaining',
            replacement: 'expect.arrayContaining',
          },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.objectContaining()',
      output: 'expect.objectContaining()',
      errors: [
        {
          messageId: 'illegalMethod',
          data: {
            method: 'jasmine.objectContaining',
            replacement: 'expect.objectContaining',
          },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.stringMatching()',
      output: 'expect.stringMatching()',
      errors: [
        {
          messageId: 'illegalMethod',
          data: {
            method: 'jasmine.stringMatching',
            replacement: 'expect.stringMatching',
          },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'jasmine.foo.any()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.foo.addMatchers()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.foo.createSpy()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'console.log(jasmine.version)',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 13, line: 1 }],
    },
    {
      code: 'jasmine.getEnv()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.empty()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.falsy()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.truthy()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.arrayWithExactContents()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.clock()',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine.MAX_PRETTY_PRINT_ARRAY_LENGTH = 42',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
    {
      code: 'jasmine["MAX_PRETTY_PRINT_ARRAY_LENGTH"] = 42',
      output: null,
      errors: [{ messageId: 'illegalJasmine', column: 1, line: 1 }],
    },
  ],
});
